Metainfo user names

Andreas-Koepke opened the pull request:
Add new endpoint for getting full name of user, supports multiple users.
Add full name of user to metadata panel.
  
Resolves #12787

Andreas-Koepke committed:
Add new endpoint for getting full name of user

Andreas-Koepke committed:
Get full name of user in metadata panel, put UID in data attribute

Andreas-Koepke committed:
Remove new endpoint - not needed

Andreas-Koepke committed:
Add user's names to data returned by getDataByIdAction

Andreas-Koepke committed:
Remove AJAX call - this.data includes names required.

Andreas-Koepke committed:
Add user's names to folder metainfo

Andreas-Koepke committed:
Update controllers to return user information

Andreas-Koepke committed:
Update all getMetaInfo methods to include users names

Andreas-Koepke commented:
This ended up being a much more significant change than expected. 

I have updated all of the associated controllers to now include the user's name. 

All of the JS files that use pimcore.element.metainfo have also been updated.

Andreas-Koepke committed:
Replace English string with translatable message

Andreas-Koepke committed:
Don't add link to user if user is unknown (deleted account).

Andreas-Koepke committed:
Merge branch 'pimcore:11.x' into metainfo_user_names

brusch commented on bundles/AdminBundle/Controller/Admin/DataObject/DataObjectController.php:
```diff
@@ -479,6 +480,24 @@ static function (Task $task) {
             $objectData['general']['versionCount'] = $objectFromDatabase->getVersionCount();
 
+            $userOwner = User::getById($objectData['general']['o_userOwner']);
+            if (empty($userOwner)) {
```
I think we should make this a bit more generic, what about adding a `getFullName()` method in the `User` object and maybe a trait that handles all the displaying related stuff?

aryaantony92 commented:
@Andreas-Koepke Thank you for the PR. And just a gentle reminder about the generic approach. Please refer to the comment above.

Andreas-Koepke commented:
Sounds good to me. I can make these changes later this week.

aryaantony92 commented:
> Sounds good to me. I can make these changes later this week.

Thank you 👍

Andreas-Koepke committed:
Merge branch 'pimcore:11.x' into metainfo_user_names

Andreas-Koepke committed:
Add getFullName method to User object

Andreas-Koepke committed:
Fix to code - missing bracket

Andreas-Koepke commented:
I have added in the new method for the User object and updated my code.

As for the trait for the display related code, I am not sure exactly what part of my changes you are referring to.

aryaantony92 commented on bundles/AdminBundle/Controller/Admin/Asset/AssetController.php:
```diff
@@ -217,6 +218,24 @@ static function (Task $task) {
             $asset->getScheduledTasks()
         );
 
+        $userOwner = User::getById($asset->getUserOwner());
+        if (empty($userOwner)) {
+            $data['userOwnerUsername'] = '';
+            $data['userOwnerFullname'] = $this->trans('user_unknown');
+        } else {
+            $data['userOwnerUsername'] = $userOwner->getName();
+            $data['userOwnerFullname'] = $userOwner->getFullName();
+        }
+
+        $userModification = ($asset->getUserOwner() == $asset->getUserModification()) ? $userOwner : User::getById($asset->getUserModification());
+        if (empty($userModification)) {
+            $data['userModificationUsername'] = '';
+            $data['userModificationFullname'] = $this->trans('user_unknown');
+        } else {
+            $data['userModificationUsername'] = $userModification->getName();
+            $data['userModificationFullname'] = $userModification->getFullName();
+        }
+
```
We have similar code (for username and usermodification) in DataObjectController, DocumentController etc. 
If we add this code in a trait, then different controllers can access it from there. Thanks!

Andreas-Koepke replied:
The new trait has been added, please review the updated code.

I have also made a fix to the metainfo panel so it handles the system account properly.

aryaantony92 replied:
Thank you! . Could you please resolve the conflicts also ?

Andreas-Koepke committed:
Add trait for Controllers to get user names

Andreas-Koepke committed:
Handle system user account in metainfo panel

Andreas-Koepke committed:
Merge branch 'metainfo_user_names_update' of github.com:Andreas-Koepke/pimcore into metainfo_user_names

Andreas-Koepke committed:
Remove use statements no longer required.
Move UserNameTrait file to new location.

Andreas-Koepke committed:
Update getFolderAction in DataObjectController to use new trait.

Andreas-Koepke committed:
Remove unnecessary whitespace

blankse commented on bundles/AdminBundle/src/Controller/Traits/UserNameTrait.php:
```diff
@@ -0,0 +1,45 @@
+ */
+trait UserNameTrait
+{
+    /**
+     * @param int $user_id The User ID.
+     */
+    protected function getUserName($user_id) {
```
```suggestion
    /**
     * @param int $user_id The User ID.
     *
     * @return array{username: string, fullname: string}
     */
    protected function getUserName(int $user_id): array
    {
```

We should use type hints for new methods

blankse commented on models/User.php:
```diff
@@ -216,6 +216,14 @@ public function setLastname($lastname)
         return $this;
     }
 
+    /**
+     * @return string|null
+     */
+    public function getFullName()
```
```suggestion
    public function getFullName(): string
```
We should use type hints for new methods

blankse commented on bundles/AdminBundle/src/Controller/Admin/Document/DocumentControllerBase.php:
```diff
@@ -465,4 +468,18 @@ protected function saveDocument(Model\Document $document, Request $request, bool
         return [$task, $document, $version];
     }
+
+    /**
+     * @param Model\Document $document
+     * @param array $data
+     */
```
```suggestion
```
We should avoid redundant php docs

Andreas-Koepke commented:
Merge conflicts have been resolved and some final modifications made.

aryaantony92 commented:
> Merge conflicts have been resolved and some final modifications made.

Thank you.

NiklasBr commented on bundles/AdminBundle/src/Controller/Traits/UserNameTrait.php:
```diff
@@ -0,0 +1,45 @@
+
+/**
+ * @internal
+ */
```
Why `internal`? Seems like a useful Trait which could be used elsewhere.

Andreas-Koepke replied:
All of the other traits in that folder were also marked as internal so I just made it the same.

brusch replied:
Agree, we should keep that one `@internal`.

Andreas-Koepke committed:
Update bundles/AdminBundle/src/Controller/Traits/UserNameTrait.php

Co-authored-by: Sebastian Blank <sebastian.bl@gmx.de>

Andreas-Koepke committed:
Update models/User.php

Co-authored-by: Sebastian Blank <sebastian.bl@gmx.de>

aryaantony92 commented on bundles/AdminBundle/src/Controller/Traits/UserNameTrait.php:
```diff
@@ -0,0 +1,48 @@
+     *
+     * @return array{username: string, fullname: string}
+     */
+    protected function getUserName(int $user_id): array
```
```suggestion
    protected function getUserName(int $userId): array
```

aryaantony92 commented on bundles/AdminBundle/src/Controller/Traits/UserNameTrait.php:
```diff
@@ -0,0 +1,48 @@
+     */
+    protected function getUserName(int $user_id): array
+    {
+        $user = User::getById($user_id);
```
```suggestion
        $user = User::getById($userId);
```

aryaantony92 commented on bundles/AdminBundle/src/Controller/Traits/UserNameTrait.php:
```diff
@@ -0,0 +1,48 @@
+        if (empty($user)) {
+            $data = [
+                'username' => '',
+                'fullname' => $this->trans('user_unknown'),
```
```suggestion
                'userName' => '',
                'fullName' => $this->trans('user_unknown'),
```

aryaantony92 replied:
Could you please change the 'username' and 'fullname' also to Camel case like the suggestion ? Thanks

Andreas-Koepke replied:
I have made this change and the associate changes to the controllers.

Andreas-Koepke committed:
Update bundles/AdminBundle/src/Controller/Traits/UserNameTrait.php

Co-authored-by: aryaantony92 <97134765+aryaantony92@users.noreply.github.com>

Andreas-Koepke committed:
Update bundles/AdminBundle/src/Controller/Traits/UserNameTrait.php

Co-authored-by: aryaantony92 <97134765+aryaantony92@users.noreply.github.com>

Andreas-Koepke committed:
Update bundles/AdminBundle/src/Controller/Admin/Document/DocumentControllerBase.php

Co-authored-by: Sebastian Blank <sebastian.bl@gmx.de>

Andreas-Koepke committed:
Fix function doc after variable change

Andreas-Koepke committed:
Change case of array keys in UserNameTrait

Andreas-Koepke committed:
Update documentation

aryaantony92 commented:
@Andreas-Koepke I just tested the PR and it's working fine. But, what about moving the conditions like https://github.com/pimcore/pimcore/blob/a4fd44d71f805aa7e48f5475e76adcbcaff97d24/bundles/AdminBundle/public/js/pimcore/asset/asset.js#L495
https://github.com/pimcore/pimcore/blob/a4fd44d71f805aa7e48f5475e76adcbcaff97d24/bundles/AdminBundle/public/js/pimcore/asset/asset.js#L496 in Js files to the controller ? So, we need to pass only one variable from the controller (if full name exists, then full name otherwise the username). Since, we are using either full name/username in the js files, we don't need both of them in the js.

Andreas-Koepke committed:
Put username into fullname field if there is no name provided.
Simplify code in metainfo view

Andreas-Koepke commented:
I just made a change to the UserNameTrait that checks if the user's full name is available. I have removed the check from JS files as it is no longer necessary.

At the moment, I have left both variables in the data returned by the controllers.

Andreas-Koepke committed:
Rename key in metainfo data to just usermodification_name or userowner_name.

aryaantony92 commented:
@Andreas-Koepke Thanks a lot!
